New Google LiteRT export replaces legacy TFLite and TF.js

ambitious-octopus opened the pull request:
This pull request adds support for exporting models to the `liteRT` format in the Ultralytics framework. The changes cover updates to export format definitions, the exporter logic, and the backend initialization to ensure `liteRT` is recognized and handled correctly.

### Exporter enhancements

* Added `liteRT` to the list of supported export formats in `export_formats()` within `ultralytics/engine/exporter.py`, including its arguments and capabilities.
* Updated the exporter logic to handle the `litert` flag, ensuring the format is recognized and processed during export.
* Implemented the `export_litert` method, which uses the `ai_edge_torch` library to convert and export models to the `liteRT` format, and saves associated metadata.
* Integrated the new export method into the exporter’s workflow, so that when the `litert` flag is set, the model is exported using the new method.

### Backend integration

* Updated the backend initialization in `ultralytics/nn/autobackend.py` to recognize the `litert` format, ensuring compatibility throughout the model loading and inference pipeline.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary
This PR introduces **LiteRT** as Ultralytics’ new unified `.tflite` export and inference path, replacing standalone **TFLite** and **TF.js** workflows across the codebase, docs, tests, and platform references 🚀

### 📊 Key Changes
- **Added a new `litert` export format** for YOLO models that outputs a single `.tflite` file for **mobile, edge, embedded, and browser** deployment.
- **Deprecated `tflite` and `tfjs` export formats**:
  - `format="tflite"` and `format="tfjs"` still work, but now show a warning and automatically export as `format="litert"`.
- **Added a new LiteRT inference backend**:
  - `.tflite` models now load through a dedicated `LiteRTBackend` in `AutoBackend`.
  - Includes support for both newer LiteRT exports and older legacy `.tflite` files.
- **Expanded quantization support** for LiteRT:
  - Added support for `w8a32` quantization.
  - Updated validation and export rules to reflect LiteRT-specific quantization behavior.
- **Updated export system and dependencies**:
  - New `export-litert` dependency group added in `pyproject.toml`.
  - Export format tables and logic now list **LiteRT** instead of separate TFLite/TF.js entries.
- **Added tests for LiteRT export** ✅
  - New test coverage checks LiteRT export across tasks and quantization modes.
  - Old TFLite-specific export tests were removed.
- **Large documentation refresh** 📚
  - Added a full new [LiteRT integration guide](../integrations/litert.md).
  - Marked **TFLite** and **TF.js** docs as deprecated.
  - Replaced many references across guides, model pages, platform docs, and FAQs from TFLite/TF.js to LiteRT.
- **Cleaned up old examples**
  - Removed the older `YOLOv8-TFLite-Python` example and related references.

### 🎯 Purpose & Impact
- **Simplifies deployment** 🌍  
  Users now have **one export target** for Android, iOS, edge devices, and browser use cases instead of juggling separate TFLite and TF.js paths.
- **Reduces confusion** 🧭  
  The docs, platform UI/docs, and export tables now reflect Google’s newer **LiteRT** naming and workflow more consistently.
- **Improves backward compatibility** 🔄  
  Existing users can still use `tflite` or `tfjs` commands for now, easing migration without breaking workflows immediately.
- **Strengthens on-device support** 📱  
  A dedicated LiteRT backend and updated quantization options make `.tflite` deployment more capable and better aligned with modern edge/browser runtimes.
- **Main impact is on export/deployment workflows, not model accuracy or training**  
  This PR is mostly about **deployment unification, documentation clarity, and runtime support**, so most users will mainly notice cleaner exports and updated terminology.

ambitious-octopus committed:
added initial tests

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Address the LiteRT exporter’s fixed sample tensor so it matches each model’s actual input shape, and save `metadata.yaml` beside the generated `.tflite` to keep artifacts consistent and avoid overwrites.

💬 Posted 2 inline comments

onuralpszr committed:
Merge branch 'main' into litert-export

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
docs: Add docstring for export_litert method to clarify YOLO model export process

UltralyticsAssistant committed:
Merge branch 'main' into litert-export

glenn-jocher committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

glenn-jocher committed:
Merge branch 'main' into litert-export

lakshanthad committed:
Merge branch 'main' into litert-export

Signed-off-by: Lakshantha Dissanayake <lakshantha@ultralytics.com>

UltralyticsAssistant committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

UltralyticsAssistant committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
fixed export and inference

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
fix: update liteRT export to support FP16/INT8 quantization and improve interpreter loading

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

ambitious-octopus committed:
Merge 1c4da3df36d3917c0a57d74a6babd889bbb21c93 into ee2ac9e43491e5ca61a158fb3a42e621a6710ee1

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

ambitious-octopus committed:
Merge 51b13bb1b5af2b2984d6340f711dfa825a0172d5 into ee2ac9e43491e5ca61a158fb3a42e621a6710ee1

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

lakshanthad committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

ambitious-octopus reviewed (commented):
@onuralpszr @lakshanthad testing with:

```python
from ultralytics import YOLO

# Load model
model = YOLO("yolo26n.pt")
del model.model.model[-1].one2one_cv2


def test_export_and_val(export_args, description):
    """Test export with given args and validate."""
    print(f"\n{'='*60}")
    print(f"Testing: {description}")
    print(f"Export args: {export_args}")
    print("=" * 60)

    # Export
    print("\nExporting...")
    
    exported_path = model.export(format="litert", imgsz=640, data="coco128.yaml", **export_args)
    print(f"Exported to: {exported_path}")

    # Validate
    print("\nValidating...")
    litert_model = YOLO(exported_path, task="detect")
    results = litert_model.val(data="coco128.yaml", imgsz=640)
    print(f"mAP50-95: {results.box.map:.4f}")
    return results.box.map


# Test cases
print("\n" + "=" * 60)
print("LiteRT Export Tests with Quantization")
print("=" * 60)

# Test 1: FP32 (default)
mAP_fp32 = test_export_and_val({}, "FP32 (default)")

# Test 2: FP16 (half)
mAP_fp16 = test_export_and_val({"half": True}, "FP16 (half precision)")

# Test 3: INT8
mAP_int8 = test_export_and_val({"int8": True}, "INT8 quantization")

# Summary
print("\n" + "=" * 60)
print("SUMMARY")
print("=" * 60)
print(f"FP32 mAP50-95: {mAP_fp32:.4f}")
print(f"FP16 mAP50-95: {mAP_fp16:.4f}")
print(f"INT8 mAP50-95: {mAP_int8:.4f}")
```

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
feat: ✨ Adding better file naming similar to current tflite export

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
docs: 📝 documentation for litert export in exporter.py

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

Y-T-G commented:
Is it now called `litert-torch` instead of `ai-edge-torch`?

https://pypi.org/project/litert-torch/

ambitious-octopus committed:
Merge branch 'main' into litert-export

ambitious-octopus reviewed (commented):
Hey @itsMarco-G, here is the PR.

onuralpszr committed:
chore(exporter): 📦 update requirement from ai-edge-torch-nightly to ai-edge-torch stable

Signed-off-by: Onuralp SEZER <thunderbirdtr@gmail.com>

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
refactor: ♻ export_litert method to use litert_torch for YOLO model export and update logging information

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Merge branch 'main' into litert-export

lakshanthad committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

lakshanthad committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
added nighly

ambitious-octopus committed:
merged

ambitious-octopus committed:
Merge 33843ef3b2a0332fa18ae4f12a5bcb3c778d62d5 into ad8858db65df71624d20396ca3ce301cbba29f05

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

ambitious-octopus committed:
added litert backend

ambitious-octopus committed:
Merge c49bfe1d4553fac629a6b722f0ced35736c94b7e into ad8858db65df71624d20396ca3ce301cbba29f05

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

lakshanthad committed:
Merge branch 'main' into litert-export

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
--amend

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
refactor: LiteRT export utilities and integrate into export folder

onuralpszr committed:
fix: update requirement for LiteRT model loading to ai-edge-litert-nightly

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

lakshanthad committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

Kanna-Sridhar commented:
[embedding_output.zip](https://github.com/user-attachments/files/26955549/embedding_output.zip)
Hi,
I used the Litert export format for the yolo26n model and I was able to extract the layer embedding for layer_0 with tensor shape [1,320,320,16] which is in the dropped file above. so how to verify these values are correct or to be more specific what do those values mean??

lakshanthad committed:
Merge branch 'main' into litert-export

glenn-jocher commented:
Those values are intermediate feature activations, not final detections: in LiteRT the tensor is usually `NHWC`, so your `[1,320,320,16]` matches PyTorch `[1,16,320,320]` from an early feature layer. The best verification is to run the exact same preprocessed input through the original PyTorch model, capture the matching layer output, transpose it to `NHWC`, and compare with a small tolerance; exact equality is not expected, especially for quantized exports. The [exporter reference](https://docs.ultralytics.com/reference/engine/exporter/) and [TFLite export guide](https://docs.ultralytics.com/integrations/tflite/) should help if you want to trace the exported tensors further.

onuralpszr committed:
fix: update LiteRT export to remove FP16 option and clarify INT8 quantization

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
fix: update LiteRT export to support FP16 quantization

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

Kanna-Sridhar commented:
[test.py](https://github.com/user-attachments/files/27230201/test.py)
Hi,
The uploaded test file displays the MSE error rate between pytorch and litert outputs which is about 24.3393 and below is the snap of that so does this sound correct??
<img width="880" height="234" alt="image" src="https://github.com/user-attachments/assets/7e78cdbe-b7e3-4df4-b507-a87eed444ccc" />
<

glenn-jocher commented:
Not really—if this is intended to be a like-for-like layer comparison, `MSE=24.3393` usually points to a mismatch rather than a correct match, most often `NCHW` vs `NHWC`, different preprocessing, or comparing quantized values without dequantization; for `.tflite` outputs, compare the exact same input and layer after applying the tensor scale/zero-point, and the [TensorFlow export reference](https://docs.ultralytics.com/reference/utils/export/tensorflow/) is a useful hint here since the calibration path uses `BHWC` images.

UltralyticsAssistant committed:
Merge branch 'main' into litert-export

Kanna-Sridhar commented:
Hi,
i have attached images for  the pytorch and litert embedding for layer 0 since extracted file size is more than 25mb for the yolo26n model with code snippet so please whether the layer embeddings are correct
<img width="1028" height="572" alt="Screenshot 2026-05-05 120119" src="https://github.com/user-attachments/assets/ba0b6b69-ea6a-45f4-aa74-8af4b29b3f14" />
<img width="1092" height="669" alt="Screenshot 2026-05-05 120322" src="https://github.com/user-attachments/assets/feb66989-628d-4b2b-a0fb-86810dd91fcb" />
[test (1).py](https://github.com/user-attachments/files/27388763/test.1.py)

glenn-jocher commented:
From screenshots alone I can’t reliably confirm the embeddings are correct; please paste the text output for the LiteRT tensor `shape`, `dtype`, and `quantization` parameters plus the exact PyTorch hook layer, because the check needs to be done on the same input after dequantizing LiteRT outputs and aligning `NCHW`/`NHWC`, and the [TensorFlow export reference](https://docs.ultralytics.com/reference/utils/export/tensorflow/) is the relevant guide here.

Kanna-Sridhar commented:
but the file size are more than 25 MB and compressed to .tar format but i couldnt drop it here

glenn-jocher commented:
No worries—you don’t need the full archive; please share a minimal repro instead with one test image, the exact PyTorch hook layer name, the text output of `interpreter.get_output_details()`, and a tiny tensor slice like `output[0, :4, :4, :4]` from both PyTorch and dequantized LiteRT, which is enough for us to verify whether the embeddings align.

Kanna-Sridhar commented:
[pytorch slice and litert slice.txt](https://github.com/user-attachments/files/27465521/pytorch.slice.and.litert.slice.txt)
[test (1).py](https://github.com/user-attachments/files/27465523/test.1.py)
i have uploaded the tensor slice for pytorch and litert also the image used for this is bus.jpg and model used is yolo26n.pt so hence test(1).py file contains every details how these values were extracted because i feel like the values are close but not identical

glenn-jocher commented:
Yes, that’s expected—LiteRT intermediate activations should be numerically close but not bit-identical to PyTorch, so the more reliable validation is to compare the final predictions on the same image after any dequantization; if the boxes, scores, and classes remain close, the exported layer output is behaving as expected.

Kanna-Sridhar commented:
Thank you @glenn-jocher for the support

glenn-jocher commented:
You’re welcome, glad it helped; if anything else comes up with the LiteRT export or output validation, feel free to post the details here.

ambitious-octopus committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge a38f0e97921fc1066158ce2b3558f6ce153028e1 into 96349afb2a4600121a6645d7eae08ec30f1af386

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

ambitious-octopus committed:
added quantization

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
fixed end2end

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
Merge branch 'main' into litert-export

Signed-off-by: Francesco Mattioli <Francesco.mttl@gmail.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
CHECKPOINT COMMIT: end2end

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Merge branch 'main' into litert-export

ambitious-octopus committed:
fixed end2end

onuralpszr committed:
LiteRT: fix INT8 export accuracy via coord normalization, drop FP16 export

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
feat: add LiteRT support in exporter and improve denormalization logic

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
fix: disable end2end support for LiteRT INT8 export and improve denormalization handling

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge remote-tracking branch 'origin/main' into litert-export

# Conflicts:
#	ultralytics/engine/exporter.py
#	ultralytics/nn/autobackend.py

onuralpszr committed:
Drop legacy tflite and tfjs export formats (covered by LiteRT)

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Add unified LiteRT integration docs, remove legacy tflite/tfjs pages

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Add LiteRT row to notebook export-formats table

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

onuralpszr committed:
Benchmark LiteRT as first-class format via export-litert extra

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Fix docs to use LiteRT/LiteRT.js for mobile and web deployment

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-25 12:21:13 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28169625828): `ci.yml` run 24254
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28169627097): `docker.yml` run 2416

onuralpszr committed:
Merge remote-tracking branch 'origin/main' into litert-export

# Conflicts:
#	docs/en/guides/optimizing-openvino-latency-vs-throughput-modes.md
#	docs/en/help/FAQ.md
#	docs/en/integrations/rockchip-rknn.md
#	docs/en/integrations/tf-savedmodel.md
#	docs/en/integrations/tfjs.md
#	docs/en/integrations/tflite.md

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Sort TASKS in litert matrix test for deterministic xdist collection

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-25 12:51:45 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28171332173): `ci.yml` run 24267
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28171333435): `docker.yml` run 2417

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge remote-tracking branch 'origin/main' into litert-export

# Conflicts:
#	docs/en/guides/defining-project-goals.md

onuralpszr committed:
Merge branch 'main' into litert-export

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Replace tflite/tfjs with LiteRT in deployment options to fix docs build

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-25 13:43:37 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28174512796): `ci.yml` run 24301
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28174514135): `docker.yml` run 2421

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge branch 'main' into litert-export

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

Laughing-q committed:
Merge branch 'main' into litert-export

Laughing-q committed:
Updates

Laughing-q commented:
@ambitious-octopus @onuralpszr Guys I found a couple of issue here:
- new litert does not support `nms` here like `tfite` does, running with `yolo export model=yolo11n.pt nms format=litert` and run inference with exported model it'd get incorrect results. Seeing this as a regression.
- new `LiteRTBackend` is not compatible with old exported as `tflite` weight:
  ```bash
  # export using main branch
  yolo export model=yolo26n.pt format=tflite
  # running on this branch
  yolo predict task=detect model=yolo26n_saved_model/yolo26n_float32.tflite imgsz=640
  ```
  would raise error, dimension mismatch, which I guess it's from the discrepancy of the input dimention between old tflite and litert, NCHW vs NHWC.
  Even if I temporarily update the code to bypass this dimension error it can run smoothly but the inference results are incorrect, I can tell the predict cls and conf scores are correct but bounding box coordinates are incorrect. I think this might be related to the discrepancy of normalize/denormalize output between tflite and litert. See the visual below:
<img width="810" height="1080" alt="bus" src="https://github.com/user-attachments/assets/e88f0eea-a93b-4061-8a72-67e8baf8976a" />

- Also introducing the new litert seems not eliminate the existence of onnx2tf package since we still need it for `pb` and `edgetpu`? 
  - https://github.com/ultralytics/ultralytics/blob/ab45ece7bbccbd7c03cc6a86dd82c612b75bf056/ultralytics/engine/exporter.py#L725-L729
  - https://github.com/ultralytics/ultralytics/blob/6c53a87e817310d0054d2129d388f5ca383e85fe/ultralytics/utils/export/tensorflow.py#L102-L115
  - which means we are not eliminating dependencies, but introducing new one?

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr commented:
> * Also introducing the new litert seems not eliminate the existence of onnx2tf package since we still need it for `pb` and `edgetpu`?

Yes because of tensorflow I cannot remove it but litert bypass those deps and make export easier for tflite.

onuralpszr commented:
> * new litert does not support `nms` here like `tfite` does, running with `yolo export model=yolo11n.pt nms format=litert` and run inference with exported model it'd get incorrect results. Seeing this as a regression.

I checked nms but new litert doesn't support torchvision.nms so I couldn't export this but I will check again

onuralpszr commented:
> * would raise error, dimension mismatch, which I guess it's from the discrepancy of the input dimention between old tflite and litert, NCHW vs NHWC.
>   Even if I temporarily update the code to bypass this dimension error it can run smoothly but the inference results are incorrect, I can tell the predict cls and conf scores are correct but bounding box coordinates are incorrect. I think this might be related to the discrepancy of normalize/denormalize output between tflite and litert. See the visual below:

I will re-check and make sure okay with older style

onuralpszr committed:
update LiteRT export support for macOS in documentation and code

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-26 11:04:13 UTC with `@ultralytics/run-ci` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28233990989): `ci.yml` run 24379

onuralpszr committed:
Support legacy onnx2tf TFLite (NHWC) in LiteRTBackend

Detect input layout from the interpreter and permute NHWC inputs so old
onnx2tf .tflite exports load through LiteRTBackend. Outputs share the same
layout for both export paths, so detection and proto handling are unchanged.
Verified old and new produce identical detections across detect, segment,
pose, obb, classify.

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Allow LiteRT benchmark on macOS

Align the benchmark platform guard with the export guard, which already
permits macOS.

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-26 11:19:54 UTC with `@ultralytics/run-ci` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28234727820): `ci.yml` run 24381

onuralpszr committed:
Denormalize legacy end2end TFLite output in LiteRTBackend

Legacy onnx2tf TFLite normalizes even the end2end (post-NMS) output, while
litert-torch outputs pixel coordinates. Denormalize the end2end output by
image size only for legacy NHWC models, so end2end exports such as yolo26n
predict correct boxes instead of collapsing near the origin. Verified old and
new produce identical detections across detect, segment, pose, obb.

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Merge main and migrate LiteRT export to the quantize argument

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Add LiteRT dynamic INT8 w8a32 quantization scheme

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Add LiteRT w8a16 INT8 weight INT16 activation quantization scheme

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

onuralpszr committed:
Enable classify INT8 and w8a16 in LiteRT export test matrix

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Skip LiteRT benchmark on macOS GitHub Actions to avoid abseil mutex deadlock

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-29 12:52:32 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28373325604): `ci.yml` run 24741
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28373326966): `docker.yml` run 2473

onuralpszr committed:
Embed metadata in tflite for single file LiteRT export

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Share tflite metadata reader and fix stale LiteRT references

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

onuralpszr committed:
Drop metadata yaml fallback from LiteRT backend for single file tflite

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

onuralpszr committed:
Merge branch 'main' into litert-export

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Load LiteRT from single tflite file and drop unused directory handling

Signed-off-by: Onuralp SEZER <onuralp@ultralytics.com>

glenn-jocher committed:
Merge branch 'main' into litert-export

onuralpszr commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-29 15:54:03 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28384967295): `ci.yml` run 24774
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28384968698): `docker.yml` run 2478

glenn-jocher committed:
Align LiteRT PR with Core Principles

- Remove orphaned tensorflowjs dep + its setuptools<=81 pin (tfjs export removed)
- LiteRTBackend: restore semantic class-map guard and NHWC->NCHW transpose for
  legacy onnx2tf semantic logits and segment protos
- LiteRTBackend: derive pose keypoint start from 4+nc to match _NormalizeCoords
  (was hardcoded for single-class pose)
- torch2litert: type calibration_dataset/metadata as Optional to match callers
- Fix stale TFLite->LiteRT comment; add fraction arg to LiteRT export-table row

glenn-jocher committed:
Bump version to 8.4.83 for LiteRT export release

glenn-jocher committed:
Add w8a32 to quantize help text in default.yaml

glenn-jocher committed:
Keep float I/O for static int8/w8a16 LiteRT exports

litert-torch + ai-edge-quantizer static recipes quantized the graph INPUT/OUTPUT
tensors to int8/int16, so format=litert quantize=8|w8a16 produced int8-I/O .tflite.
The Python backend masked this by (de)quantizing I/O, but on-device consumers
(yolo-flutter-app) feed/read floats and have no boundary (de)quant, so the official
int8 assets would break. Force NO_QUANTIZE on the INPUT/OUTPUT ops to restore the
historical onnx2tf fp32-in/out contract (weights/activations stay int8/int16; GPU
delegate prefers float I/O). Drop the now-dead int8 I/O (de)quant in LiteRTBackend
and assert float32 I/O in test_export_litert_matrix so the contract can't silently
regress.

glenn-jocher committed:
Use yolo26n for LiteRT export-env smoke test

glenn-jocher committed:
LiteRT: require square imgsz; restore full-integer .tflite (de)quant

- export_litert asserts square imgsz: _NormalizeCoords/backend scale coords by a
  single image size, so non-square exports silently mis-scaled the y axis.
- Restore the int8/int16 input-quantize and output-dequantize in LiteRTBackend so
  legacy onnx2tf and third-party full-integer .tflite still run (Ultralytics' own
  exports keep float I/O, asserted in test_export_litert_matrix).

glenn-jocher committed:
LiteRT: feed full-batch INT8 calibration samples matching traced batch

litert-torch traces a fixed batch (im.shape[0]); the static int8/w8a16
calibration fed singleton [1,C,H,W] samples, mismatching the graph input for
batch>1. Feed whole dataloader batches that match im's batch dim (identical for
the default batch=1).

glenn-jocher committed:
Docs: LiteRT FP16 is not a separate export (runtime via GPU delegate)

The export precision table marked LiteRT FP16 (quantize=16) as a supported
export, but litert is not in FP16_FORMATS so quantize=16 is rejected. Mark it
unsupported and keep the GPU-delegate runtime-FP16 note.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
LiteRT: run CI export env on Python 3.13; guard empty INT8 calibration

- EXPORT_ENVS litert -> Python 3.13 (litert-torch/ai-edge stack installs, imports,
  and exports cleanly on 3.13; validated all quantize modes)
- torch2litert: raise a clear error when no calibration batch matches the traced
  batch (dataset smaller than batch) instead of an opaque empty-calibrate failure

glenn-jocher committed:
LiteRT: support non-square imgsz via per-axis coord normalization

_NormalizeCoords now divides x by width and y by height (matching the backend's
per-axis denorm) instead of one scalar, so non-square exports are correct; drop
the square-imgsz assert. The box divisor [W,H,W,H] is uniform-magnitude so it
stays INT8-safe. Validated: 640x480 FP32 and INT8 exports both detect bus+persons
with matching boxes.

glenn-jocher committed:
Merge branch 'main' into litert-export

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

glenn-jocher commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-29 20:50:33 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28401752401): `ci.yml` run 24848
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28401753384): `docker.yml` run 2484

glenn-jocher committed:
Accept deprecated tflite/tfjs format args as LiteRT aliases

format=tflite and format=tfjs are no longer rejected: they emit a deprecation
warning (deprecated 8.4.83, replaced by unified Google LiteRT, link to litert
docs) and export a LiteRT model. They stay out of the advertised allowlist.

glenn-jocher committed:
Restore tflite/tfjs as LiteRT deprecation docs pages (SEO retention)

Retain the TFLite and TF.js integration pages with a prominent deprecation
banner (deprecated 8.4.83, replaced by unified Google LiteRT, link to litert.md)
so inbound searches for the old formats are not lost; fix the export examples to
format=litert and re-add both to the mkdocs nav.

glenn-jocher committed:
Restore clean tutorial.ipynb export-table diff

Recover the notebook to its original array-of-strings cell format and apply only
the intended export-table change (drop TF Lite + TF.js rows, add LiteRT) instead
of the full-notebook re-serialization.

glenn-jocher committed:
Clean up tflite/tfjs deprecation pages for LiteRT consistency

- tflite.md: predict/val snippets load yolo26n.tflite (FP32 LiteRT has no _float32 suffix)
- tfjs.md: replace _web_model / TensorFlow.js-runtime deployment prose with the
  LiteRT .tflite path (local predict/val + browser via LiteRT.js)
- both: drop the nms export-arg row (LiteRT does not accept nms)

glenn-jocher committed:
Rename LiteRT w8a32 export suffix _dynamic_int8 -> _w8a32 for consistency

quantize='w8a32' now writes *_w8a32.tflite, matching quantize='w8a16' -> *_w8a16.tflite. quantize=8 keeps the established *_int8.tflite convention used by release assets.

glenn-jocher committed:
List valid quantize options in unsupported-precision export errors

Unsupported quantize values now report the precisions the target format actually accepts (e.g. format='litert' supports 8, 'w8a16', 'w8a32', 32) instead of only saying it is unsupported.

glenn-jocher committed:
Fix deprecated tflite/tfjs docs quantize options; restore yolov5 links

Alias pages route to LiteRT, so document its quantize options (8/w8a16/w8a32/32, no FP16) instead of the rejected quantize=16. Revert yolov5 model-export links to the restored tflite/tfjs deprecation pages.

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Update qnn quantize test matcher for new precision-labeled error text

glenn-jocher commented:
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-06-29 22:11:43 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/28406019414): `ci.yml` run 24863
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/28406020435): `docker.yml` run 2486

UltralyticsAssistant commented:
🎉 Huge congratulations on merging this major LiteRT unification PR!

“As we look ahead into the next century, leaders will be those who empower others.” — Bill Gates

This update does exactly that: by bringing export, inference, docs, tests, and platform references together under **LiteRT**, it makes YOLO deployment simpler, clearer, and more accessible across mobile, edge, embedded, and browser workflows. The thoughtful backward compatibility and expanded quantization support make the impact even stronger.

Big thanks to @ambitious-octopus for driving this forward, with valuable contributions from @onuralpszr, @Y-T-G, @Kanna-Sridhar, @glenn-jocher, @lakshanthad, and @Laughing-q. This is a meaningful step toward a more unified and user-friendly Ultralytics experience 🚀
